spec/util/e2e-test.ts

```diff
@@ -0,0 +1,256 @@
+                }]
+            },
+            url: "not-used",
+        }));
+        console.log('Starting bridge');
+        await this.ircBridge.run(null);
+    }
+
+    public async tearDown(): Promise<void> {
+        if (this.dbPath) {
+            await rm(this.dbPath, { recursive: true });
+        }
+        await Promise.allSettled([
+            this.dbPath && rm(this.dbPath, { recursive: true }),
+            this.ircBridge?.kill(),
+            super.tearDown(),
+            this.homeserver?.users.map(c => c.client.stop()),
+            this.homeserver && destroyHS(this.homeserver.id),
+        ]);
+    }
+}
+
+export class E2ETestMatrixClient extends MatrixClient {
+
+    public async waitForRoomEvent(
+        opts: {eventType: string, sender: string, roomId?: string, stateKey?: string}
+    ): Promise<{roomId: string, data: unknown}> {
+        const {eventType, sender, roomId, stateKey} = opts;
+        return this.waitForEvent('room.event', (eventRoomId: string, eventData: {
+            sender: string, type: string, state_key?: string, content: unknown
+        }) => {
+            console.info(`Got ${eventRoomId}`, eventData);
+            if (eventData.sender !== sender) {
+                return undefined;
+            }
+            if (eventData.type !== eventType) {
+                return undefined;
+            }
+            if (roomId && eventRoomId !== roomId) {
+                return undefined;
+            }
+            if (stateKey && eventData.state_key !== stateKey) {
+                return undefined;
+            }
+            return {roomId: eventRoomId, data: eventData};
+        }, `Timed out waiting for ${eventType} from ${sender} in ${roomId || "any room"}`)
+    }
+
+    public async waitForRoomInvite(
+        opts: {sender: string, roomId?: string}
+    ): Promise<{roomId: string, data: unknown}> {
+        const {sender, roomId} = opts;
+        return this.waitForEvent('room.invite', (eventRoomId: string, eventData: {
+            sender: string
+        }) => {
+            const inviteSender = eventData.sender;
+            console.info(`Got invite to ${eventRoomId} from ${inviteSender}`);
+            if (eventData.sender !== sender) {
+                return undefined;
+            }
+            if (roomId && eventRoomId !== roomId) {
+                return undefined;
+            }
+            return {roomId: eventRoomId, data: eventData};
+        }, `Timed out waiting for invite to ${roomId || "any room"} from ${sender}`)
+    }
+
+    public async waitForEvent<T>(
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        emitterType: string, filterFn: (...args: any[]) => T|undefined, timeoutMsg: string)
+    : Promise<T> {
+        return new Promise((resolve, reject) => {
+            // eslint-disable-next-line prefer-const
+            let timer: NodeJS.Timeout;
+            const fn = (...args: any[]) => {
```

Comment:
If it doesn't make TypeScript cry, `unknown` would be better.

```suggestion
        emitterType: string, filterFn: (...args: unknown[]) => T|undefined, timeoutMsg: string)
    : Promise<T> {
        return new Promise((resolve, reject) => {
            // eslint-disable-next-line prefer-const
            let timer: NodeJS.Timeout;
            const fn = (...args: unknown[]) => {
```